src/json.c

```diff
@@ -0,0 +1,411 @@
+		.size = 0,
+		.type = json_string,
+		.code = ENOSYS
+	};
+}
+
+/**
+ * @brief Parse a negative number.
+ *
+ * On entry, `arg.ptr` points to the '-' character.
+ *
+ * @param arg JSON argument.
+ * @return JSON return value.
+ */
+static force_inline struct json_ret
+json_parse_neg (struct json_arg arg)
+{
+	struct json_ret ret = {
+		.size = 1,
+		.type = json_number,
+		.code = 0
+	};
+
+	uint8_t const *p = &arg.ptr[1];
+	if (p >= arg.end) {
+		// last character is '-'
+		ret.code = ENODATA;
+		return ret;
+	}
+
+	uint8_t flags = json_lut[*p];
+
+	if (!(flags & json_dig)) {
+		// non-digit follows '-'
+		ret.code = EILSEQ;
+		return ret;
+	}
+
+	++p;
+
+	if (flags & json_1_9)
+		// -[1-9][0-9]* (instead of -0)
+		p = json_skip_digits(p, arg.end);
+
+	ret = json_parse_frc_exp((struct json_arg){
+		.ptr = p,
+		.end = arg.end,
+		.ctx = arg.ctx,
+	});
+	if (!(flags & json_1_9)) {
+		if (ret.size == 0 && p < arg.end
+		    && (json_lut[*p] & json_dig))
+			// digit follows "-0"
+			ret.code = EILSEQ;
+	}
+	ret.size += (uint64_t)(ptrdiff_t)(p - arg.ptr);
+	return ret;
+}
+
+struct json_ret
+parse_false (struct json_arg arg)
+{
+	uint8_t const *p = arg.ptr;
+	struct json_ret ret = {
+		.size = 0,
+		.type = json_false,
+		.code = 0
+	};
+
+	do {
+		if ((ptrdiff_t)(arg.end - arg.ptr) < 5) {
+			if (!(++p < arg.end && (*p != 'a' ||
+			     (++p < arg.end && (*p != 'l' ||
+			     (++p < arg.end && (*p != 's' ||
+			      ++p < arg.end))))))) {
+				ret.code = ENODATA;
+				break;
+			}
+
+		} else if (*++p == 'a' &&
+		           *++p == 'l' &&
+		           *++p == 's' &&
+		           *++p == 'e') {
+			++p;
+			break;
+		}
+
+		ret.code = *p ? EILSEQ : ENODATA;
+	} while (0);
+
+	ret.size = (uint64_t)(ptrdiff_t)(p - arg.ptr);
+	return ret;
+}
+
+struct json_ret
+parse_null (struct json_arg arg)
+{
+	uint8_t const *p = arg.ptr;
+	struct json_ret ret = {
+		.size = 0,
+		.type = json_null,
+		.code = 0
+	};
+
+	do {
+		if ((ptrdiff_t)(arg.end - arg.ptr) < 4) {
+			if (!(++p < arg.end && (*p != 'u' ||
+			     (++p < arg.end && (*p != 'l' ||
+			      ++p < arg.end))))) {
+				ret.code = ENODATA;
+				break;
+			}
+
+		} else if (*++p == 'u' &&
+		           *++p == 'l' &&
+		           *++p == 'l') {
+			++p;
+			break;
+		}
+
+		ret.code = *p ? EILSEQ : ENODATA;
+	} while (0);
+
+	ret.size = (uint64_t)(ptrdiff_t)(p - arg.ptr);
+	return ret;
+}
+
+struct json_ret
+parse_true (struct json_arg arg)
+{
+	uint8_t const *p = arg.ptr;
+	struct json_ret ret = {
+		.size = 0,
+		.type = json_true,
+		.code = 0
+	};
+
+	do {
+		if ((ptrdiff_t)(arg.end - arg.ptr) < 4) {
+			if (!(++p < arg.end && (*p != 'r' ||
+			     (++p < arg.end && (*p != 'u' ||
+			      ++p < arg.end))))) {
+				ret.code = ENODATA;
+				break;
+			}
+
+		} else if (*++p == 'r' &&
+		           *++p == 'u' &&
+		           *++p == 'e') {
+			++p;
+			break;
+		}
+
+		ret.code = *p ? EILSEQ : ENODATA;
+	} while (0);
+
+	ret.size = (uint64_t)(ptrdiff_t)(p - arg.ptr);
+	return ret;
+}
```

Comment:
Show me your implementation and I'll compare the generated binaries.

Reply:
Not the same behavior.

Reply:
Still not even close.

- `ENODATA` is the correct error code *only* if there are no invalid bytes but there are not enough bytes. You can't just slap `ENODATA` on it because the buffer has less than 4 bytes left to read. Those bytes may or may not match a prefix of `"true"`.
- You _must_ set `ret.size` to the exact number of bytes leading up to either the successful completion of a comparison, or the first invalid byte, or a premature end of buffer. The caller must know exactly at what offset an invalid byte was encountered, or at what offset the buffer came to an abrupt end. Yes, `ret.size` can only be 4 when the input contains the string `"true"`, but every other code path must set a value that makes sense, too.

I'm not sure if I'm missing something in addition to these points, but start from there.